### ISSUE TITLE
api dependancies not asyc

### ISSUE THREAD

**Issue by nathantbaker** [nathantbaker/player-count#8]
Richmond Watkins [1:31 AM] 
@nathantbaker You could run each API call in a dedicated Operation subclass and add them all to an OperationQueue. Then you could either wait for all Operations (API calls) in the OperationQueue to be finished on a background thread or you could create a separate “completion” Operation and add each API Operation as a dependency on the completion Operation so that the completion Operation would get run once all the API Operations are finished….your current approach isn’t necessarily bad but I don’t know the details of your `getGameDetails` call. If there is a path that doesn’t return a response then you run the risk of  an error preventing your callback from being executed…An OperationQueue based approach would enable you to not have to rely on an array count to determine when all calls have finished which is an improvement IMO.

@richmondwatkins cool thanks for the direction! About 1 in 5 the callback isn’t getting called. Could be one of the dependancies causing everything to fail, but i’d be nice to get back mostly complete data in that case.

John Saba [3:40 PM] 
@nathantbaker similar to the operation queue, but you may get a more terse implementation using `DispatchGroup` (https://developer.apple.com/documentation/dispatch/dispatchgroup) (edited)

[3:41] 
basically create your group, then you can use `enter`, `leave` and `notify`to count up operations, similar to how a semaphore might work but a bit easier to use.

`enter` adds a counter to the group, `leave` takes one away (you’d wanna put it inside the completion of the async op) and `notify` only dispatches to the given queue when all the enter / leave calls are matched.

After writing it out, I think your solution is more clear for what you are doing. I’ve used dispatch groups before when I needed to wait on different stuff that wasn’t all being put into the the same collection (i.e. not as cleanly checkable for an end condition). But maybe you’ll find use of this anyway.

**Comment by nathantbaker** [comment]
fixed by throttling api calls
